Policy: label CIDRs from CIDRGroups, add CIDRGroup label selector

squeed opened the pull request:
This PR adds the ability to select CIDRs by labels:

0. Stop pruning unreferenced CiliumCIDRGroups
1. Add the field `CIDRGroupSelector` to the policy CIDR rule type.
2. Add CIDRGroup labels to the ipcache prefix metadata
3. Implement CIDRGroup label selection

A follow-up PR, once this is merged, will add documentation.

The result: This CIDRGroup results in these labels:
```yaml
apiVersion: cilium.io/v2alpha1
kind: CiliumCIDRGroup
metadata:
  name: example
  labels:
    foo: bar
spec:
  externalCIDRs:
    - "1.1.1.0/24"
    - "2a02:8109:abb8:6c00::e998/128"
```

```
# cilium identity list
16777217   cidrgroup:foo=bar
           cidrgroup:io.cilium.policy.cidrgroupname/example
           reserved:world-ipv4
16777218   cidrgroup:foo=bar
           cidrgroup:io.cilium.policy.cidrgroupname/example
           reserved:world-ipv6
```

And can now be selected thus:
```yaml
apiVersion: cilium.io/v2
kind: CiliumNetworkPolicy
metadata:
  name: allow-client-to-cidrgroup-example
spec:
  endpointSelector:
    matchLabels:
      app: client
  egress:
  - toCIDRSet:
    - cidrGroupSelector:
        matchLabels:
          foo: bar
      except:
        - "1.1.1.1/32"
```

And you will see a flow like this:
```jsonc
{
  "flow": {
    "verdict": "FORWARDED",
    "IP": {
      "source": "10.244.1.212",
      "destination": "1.1.1.2",
      "ipVersion": "IPv4"
    },
    "source": { /* elided */ },
    "destination": {
      "identity": 16777217,
      "labels": [
        "cidrgroup:foo=bar",
        "cidrgroup:io.cilium.policy.cidrgroupname/example",
        "reserved:world-ipv4"
      ]
    },
  /*  more fields elided */
}


```

```release-note
Adds the ability to add labels to external CIDRs for policy selection and Hubble flows.
```

squeed committed:
Policy API: allow selecting CiliumCIDRGroup by label

This adds the API fields necessary for selecting one or more CiliumCIDRGroups
by label. A subsequent commit will add the implementation for this
field.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
policy: add all labels for CiliumCIDRGroups

This commit inserts all labels from a CIDRGroup to its prefixes in the
ipcache. It also changes the CIDRGroup pruning logic to insert all known
CIDRGroups if there are any policies that reference CIDRGroups by label.

Consider this example CIDRGroup

apiVersion: cilium.io/v2alpha1
kind: CiliumCIDRGroup
metadata:
  name: example
  labels:
    foo: bar
spec:
  externalCIDRs:
    - "1.1.1.1/32"

The IPCache will now have this metadata entry:

1.1.1.1/32 => ["cidrgroup:foo=bar", "reserved:world-ipv4",
"cidrgroup:io.cilium.policy.cidrgroupname/example"]

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
policy/api: implement CIDRGroup label selection

This adds the rule-side implementation of CIDRGroup label selection.
When a Rule selects a CiliumCIDRGroup by labels, this generates the
appropriate EndpointSelector.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

joamaki reviewed (approved):
sig-k8s LGTM

learnitall reviewed (commented):
Sweet! Would it be possible to extend this PR with some unit tests for the new functionality?

squeed committed:
cilium-cli: add cidrgroup-by-label test

This is identical to the cidrgroup-by-name test, but instead references
CIDRGroup by labels.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
policy/k8s: stop pruning unused CiliumCIDRGroups

Before this change, we would "prune" unreferenced CiliumCIDRGroups,
keeping those CIDRs out of the IPCache if they were not selected by a
policy.

This is actually not very desirable:
* We would upsert CIDRGroup metadata if any policy, anywhere, referenced
  them, without caring if that policy applied to endpoints on the node.
* Unreferenced CiliumCIDRGroups are still useful, e.g. for Hubble flows

Given that most CiliumCIDRGroups are likely to be selected by at least
one policy, somewhere, pruning them is counterproductive.

This change removes the CiliumCIDRGroup internal refcounting mechanism,
and just always inserts them in to the ipcache.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

gandro commented on pkg/policy/api/cidr.go:
```diff
@@ -196,6 +196,8 @@ func (s CIDRRuleSlice) GetAsEndpointSelectors() EndpointSelectorSlice {
 				Key:      lbl.GetExtendedKey(),
 				Operator: slim_metav1.LabelSelectorOpExists,
 			})
+		} else if r.CIDRGroupSelector != nil {
```
nit: There is an existing unit test for this function. It would be great to have this case tested as well.

gandro commented on pkg/policy/api/rule_validation.go:
```diff
@@ -557,18 +558,30 @@ func (c CIDR) sanitize() error {
-	// Either CIDRGroupRef or Cidr is allowed
-	if len(c.CIDRGroupRef) == 0 && len(c.Cidr) == 0 {
-		return fmt.Errorf("either cidrGroupRef or cidr are required")
+	// Exactly one of CIDR, CIDRGroupRef, or CIDRGroupSelector must be set
```
nit: It might be nice to have the unit test for this function extended to cover the new logic as well

gandro reviewed (approved):
Awesome work, this looks super neat! I didn't spot anything major during the code review. The fact that we already selected CIDRGroups by a "name label" earlier this year makes this change easy to reason about.

There are a few changes that I think would benefit from unit testing, but with feature freeze looming I'm also fine to push that to a follow-up PR.
